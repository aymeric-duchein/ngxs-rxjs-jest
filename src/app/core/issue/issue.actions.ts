import {Issue} from './issue.model';

export class OpenIssueModal {
  static readonly type = '[ISSUE] open issue modal';

  constructor(public issue: Issue) {
  }
}

export class HttpIssueSuccess {
  static readonly type = '[ISSUE] Http issue success';

  constructor(public issue: Issue) {
  }
}
