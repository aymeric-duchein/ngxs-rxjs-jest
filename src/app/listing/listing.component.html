<div class="search-container">
  <app-search></app-search>
</div>
<div class="sort-container">
  <app-sort></app-sort>
</div>
<app-table></app-table>

<app-add></app-add>
