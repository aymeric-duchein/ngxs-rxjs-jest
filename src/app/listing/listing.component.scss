app-add {
  position: absolute;
  bottom: 1rem;
  right: 3rem;
}


.search-container {
  padding: 2rem 8rem 0.5rem 7rem;
}

.sort-container {
  display: flex;
  padding: 0 8rem;
  justify-content: flex-end;

  app-sort {
    margin-right: 5px;
  }
}

