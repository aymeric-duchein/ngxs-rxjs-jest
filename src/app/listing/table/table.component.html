<div>
  <app-issue *ngFor="let issue of  issues$ | async" [issue]="issue" (issueClick)="issueClick($event)"></app-issue>
</div>
