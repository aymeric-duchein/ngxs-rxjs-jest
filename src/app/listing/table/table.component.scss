app-issue {
  cursor: pointer;
}

div {
  height: calc(100vh - 206px);
  overflow: auto;
  margin-left: 0.5em;
  margin-right: 0.5em;
  padding: 0 7rem;

  &::-webkit-scrollbar {
    width: 6px;
    height: 8px;
    background: rgba(9, 30, 66, .08);
    border-radius: 3px;
  }

  &::-webkit-scrollbar-thumb {
    background-color: #bfc4ce;
    border-radius: 3px;
  }
}
