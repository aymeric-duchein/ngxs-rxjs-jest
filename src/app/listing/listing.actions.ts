import {Issue, SortKey} from '../core/issue/issue.model';

export class ListingIssueClick {
  static readonly type = '[LISTING] Issue click';

  constructor(public issue: Issue) {
  }
}

export class SearchFormUpdated {
  static readonly type = '[LISTING] Search updated';

  constructor(public search: string) {
  }
}

export class SortOptionClick {
  static readonly type = '[LISTING] Sort option click';

  constructor(public sortKey: SortKey) {
  }
}
