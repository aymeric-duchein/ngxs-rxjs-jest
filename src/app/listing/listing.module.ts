import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {ListingComponent} from './listing.component';
import {SearchComponent} from './search/search.component';
import {SortComponent} from './sort/sort.component';
import {TableComponent} from './table/table.component';
import {MatButtonModule, MatFormFieldModule, MatIconModule, MatInputModule, MatMenuModule} from '@angular/material';
import {FormsModule, ReactiveFormsModule} from '@angular/forms';
import {IssueModule} from '../shared/issue/issue.module';
import {AddModule} from '../shared/add/add.module';

@NgModule({
  declarations: [ListingComponent, SearchComponent, SortComponent, TableComponent],
  imports: [
    CommonModule,
    MatFormFieldModule,
    MatIconModule,
    MatInputModule,
    ReactiveFormsModule,
    FormsModule,
    MatButtonModule,
    MatMenuModule,
    IssueModule,
    AddModule,
  ]
})
export class ListingModule {
}
