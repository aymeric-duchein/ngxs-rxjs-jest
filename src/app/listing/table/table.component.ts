import {Component} from '@angular/core';
import {IssueState} from '../../core/issue/issue.state';
import {Observable} from 'rxjs';
import {Issue} from '../../core/issue/issue.model';
import {Select, Store} from '@ngxs/store';
import {ListingIssueClick} from '../listing.actions';

@Component({
  selector: 'app-table',
  templateUrl: './table.component.html',
  styleUrls: ['./table.component.scss']
})
export class TableComponent {

  @Select(IssueState.getFilteredAndSortedIssues) issues$: Observable<Issue[]>;

  constructor(private readonly store: Store) {
  }

  issueClick(issue: Issue) {
    this.store.dispatch(new ListingIssueClick(issue));
  }
}
