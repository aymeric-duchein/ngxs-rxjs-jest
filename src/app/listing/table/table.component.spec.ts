import {ComponentFixture, TestBed} from '@angular/core/testing';

import {TableComponent} from './table.component';
import {NgxsModule, Store} from '@ngxs/store';
import {Subject} from 'rxjs';
import {Issue} from '@src/app/core/issue/issue.model';
import {MockComponent, MockModule} from 'ng-mocks';
import {MatButtonModule, MatIconModule, MatMenuModule} from '@angular/material';
import {mockSelect} from '@ngxs-labs/testing/jest';
import {IssueState} from '@src/app/core/issue/issue.state';
import {By} from '@angular/platform-browser';
import {ListingIssueClick} from '@src/app/listing/listing.actions';
import {IssueComponent} from '@src/app/shared/issue/issue.component';
import {MOCK_DONE_ISSUE, MOCK_IN_PROGRESS_ISSUE, MOCK_TODO_ISSUE} from '@mocks/issue.mock';

describe('TableComponent', () => {
  let component: TableComponent;
  let fixture: ComponentFixture<TableComponent>;
  let store: Store;
  let getFilteredIssues$: Subject<Issue[]>;
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TableComponent, MockComponent(IssueComponent)],
      imports: [
        NgxsModule.forRoot(),
        MockModule(MatButtonModule),
        MockModule(MatMenuModule),
        MockModule(MatIconModule),
      ]
    });

    store = TestBed.get(Store);
    jest.spyOn(store, 'dispatch');

    getFilteredIssues$ = mockSelect(IssueState.getFilteredIssues as any);
    fixture = TestBed.createComponent(TableComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should display issues', () => {
    getFilteredIssues$.next([MOCK_TODO_ISSUE, MOCK_DONE_ISSUE, MOCK_IN_PROGRESS_ISSUE]);
    fixture.detectChanges();
    const issues = fixture.debugElement.queryAll(By.css('app-issue'));
    expect(issues.length).toEqual(3);
    expect(fixture).toMatchSnapshot();
  });

  test('should dispatch click action on click', () => {
    getFilteredIssues$.next([MOCK_TODO_ISSUE, MOCK_DONE_ISSUE, MOCK_IN_PROGRESS_ISSUE]);
    fixture.detectChanges();
    const issue = fixture.debugElement.query(By.css('app-issue'));
    issue.triggerEventHandler('issueClick', MOCK_TODO_ISSUE);
    fixture.detectChanges();
    expect(store.dispatch).toHaveBeenCalledWith(new ListingIssueClick(MOCK_TODO_ISSUE));
  });
});
