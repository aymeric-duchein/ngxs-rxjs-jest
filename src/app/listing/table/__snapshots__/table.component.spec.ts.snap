// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent should display issues 1`] = `
<app-table
  __issues$__selector={[Function Subject]}
  store={[Function Store]}
>
  <div>
    
    <app-issue
      ng-reflect-issue="[object Object]"
    />
    <app-issue
      ng-reflect-issue="[object Object]"
    />
    <app-issue
      ng-reflect-issue="[object Object]"
    />
  </div>
</app-table>
`;
